org/qmckl_sherman_morrison_woodbury.org

```diff
@@ -0,0 +1,1775 @@
+#+TITLE: Sherman-Morrison-Woodbury
+#+SETUPFILE: ../tools/theme.setup
+#+INCLUDE: ../tools/lib.org
+
+Low- and high-level functions that use the Sherman-Morrison and
+ Woodbury matrix inversion formulas to update the inverse of a
+ non-singualr matrix
+ 
+* Headers
+  #+begin_src elisp :noexport :results none :exports none
+(org-babel-lob-ingest "../tools/lib.org")
+#+end_src
+
+  #+begin_src c :comments link :tangle (eval c_test) :noweb yes
+#include "qmckl.h"
+#include "assert.h"
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <math.h>
+
+int main() {
+  qmckl_context context;
+  context = qmckl_context_create();
+
+  qmckl_exit_code rc;
+  #+end_src
+
+  
+* Helper Functions
+  
+Helper functions that are used by the Sherman-Morrison-Woodbury kernels. 
+These functions can only be used internally by higher level functions.
+
+** ~qmckl_slagel_splitting~
+   :PROPERTIES:
+   :Name:     qmckl_slagel_splitting
+   :CRetType: double
+   :FRetType: double precision
+   :END:
+
+   ~qmckl_slagel_splitting~ is used internally to apply a list of rank-1 updates while splitting an update if necessary.
+   In case of a split it applies the first half of the update while putting the second half in waiting queue to be applied at the end.
+  
+   For a given update $u_j$ we define a threshold value $\epsilon_j$, which is the minimum value of 
+   $1+v_j^TS^{-1}u_j$ for a non-singular matrix $S$. If $1+v_j^TS^{-1}u_j \geq \epsilon_j$, 
+   the update is applied as usual. Otherwise, $u_j$ will be redefined as $\frac{u_j}{2}$, and the other half 
+   (to be applied at the end) will be defined using vectors $\frac{u_{j'}}{2}$ and $v_{j'}^T=v_{j'}^T$.
```

Comment:
```suggestion
   the update is applied as usual. Otherwise, the current update $u_j$ will be redefined as $\frac{u_j}{2}$. The missing half-update will be delayed to the end of the algorithm; it is stored in the waiting-queue `later_updates`.
```